p2p/udp.go

```diff
@@ -0,0 +1,226 @@
+	}
+
+	return res[0], nil
+}
+
+// SendWrappedMessage is a proxy method to the sendMessageImpl. it sends a wrapped message and used within MessageServer
+func (mux *UDPMux) SendWrappedMessage(nodeID p2pcrypto.PublicKey, protocol string, payload *service.DataMsgWrapper) error {
+	return mux.sendMessageImpl(nodeID, protocol, payload)
+}
+
+// SendMessage is a proxy method to the sendMessageImpl.
+func (mux *UDPMux) SendMessage(peerPubkey p2pcrypto.PublicKey, protocol string, payload []byte) error {
+	return mux.sendMessageImpl(peerPubkey, protocol, service.DataBytes{Payload: payload})
+}
+
+// sendMessageImpl finds the peer address, wraps the message as a protocol message with p2p metadata and sends it.
+func (mux *UDPMux) sendMessageImpl(peerPubkey p2pcrypto.PublicKey, protocol string, payload service.Data) error {
+	var err error
+	var peer node.Node
+
+	peer, err = mux.localLookup(peerPubkey)
+
+	if err != nil {
+		return err
+	}
+
+	//todo: Session (maybe use cpool ?)
+
+	protomessage := &pb.ProtocolMessage{
+		Metadata: NewProtocolMessageMetadata(mux.local.PublicKey(), protocol),
+	}
+
+	switch x := payload.(type) {
+	case service.DataBytes:
+		protomessage.Data = &pb.ProtocolMessage_Payload{Payload: x.Bytes()}
+	case *service.DataMsgWrapper:
+		protomessage.Data = &pb.ProtocolMessage_Msg{Msg: &pb.MessageWrapper{Type: x.MsgType, Req: x.Req, ReqID: x.ReqID, Payload: x.Payload}}
+	case nil:
+		// The field is not set.
+	default:
+		return fmt.Errorf("protocolMsg has unexpected type %T", x)
+	}
+
+	data, err := proto.Marshal(protomessage)
+	if err != nil {
+		return fmt.Errorf("failed to encode signed message err: %v", err)
+	}
+
+	if len(data) > maxMessageSize {
+		return errors.New(fmt.Sprintf("message too big (%v bytes). max allowed size = %d bytes", len(data), maxMessageSize))
+	}
+
+	// TODO: node.address should have IP address, UDP and TCP PORT.
+	// 		 for now assuming it's the same port for both.
+
+	return mux.network.Send(peer, data)
+}
+
+type udpProtocolMessage struct {
+	meta   service.P2PMetadata
+	sender p2pcrypto.PublicKey
+	msg    []byte
+}
+
+func (upm *udpProtocolMessage) Sender() p2pcrypto.PublicKey {
+	return upm.sender
+}
+
+func (upm *udpProtocolMessage) Metadata() service.P2PMetadata {
+	return upm.meta
+}
+
+func (upm *udpProtocolMessage) Bytes() []byte {
+	return upm.msg
+}
+
+// processUDPMessage processes a udp message received and passes it to the protocol, it adds related p2p metadata.
+func (mux *UDPMux) processUDPMessage(fromaddr net.Addr, buf []byte) error {
+	msg := &pb.ProtocolMessage{}
+	err := proto.Unmarshal(buf, msg)
+	if err != nil {
+		return errors.New("could'nt deserialize message")
+	}
+	sender, err := p2pcrypto.NewPubkeyFromBytes(msg.Metadata.AuthPubkey)
+	if err != nil {
+		return errors.New("no valid pubkey in message")
+	}
+	//mux.logger.Debug("Received %v from %s at %s", sender, fromaddr)
+
+	var data service.Data
+
+	if payload := msg.GetPayload(); payload != nil {
```

Comment:
what if it's neither?

Reply:
these are the only fields we accept as part of this message. if none of them are present it's probably a malformed message. I'll add a treatment for that case.